<a href="https://colab.research.google.com/github/jazaineam1/BigData2023_2/blob/main/Cuadernos/8_NoSQL_Mongo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***No Sql y Mongodb***

## ***Universidad Central***
>## **Facultad de Ingeniería y Ciencias Básicas.**
>## ***Maestría en analítica de datos***
![Imágen1](https://www.ucentral.edu.co/themes/ucentral/img/template/Universidad%20Central.png)


>## ***Big Data.***
>## ***Docente: Antonino Zainea Maya.***



**¿Qué son las bases de datos NoSQL?**

Las bases de datos NoSQL (Not Only SQL) son sistemas de gestión de bases de datos que difieren en su enfoque con respecto a las bases de datos relacionales tradicionales. Están diseñadas para manejar datos no estructurados o semiestructurados y permiten un almacenamiento y recuperación de datos más flexible y escalable. Aquí hay cinco características para ayudar a comprender este concepto:

1. 🗂️ **Flexibilidad de Esquema:** A diferencia de las bases de datos relacionales, las bases de datos NoSQL no requieren un esquema fijo. Puedes almacenar datos en formato JSON, XML o documentos BSON sin necesidad de definir tablas y columnas.

2. 🚀 **Escalabilidad Horizontal:** Las bases de datos NoSQL son altamente escalables, lo que significa que puedes distribuir fácilmente tu base de datos en múltiples servidores para manejar grandes volúmenes de datos y cargas de trabajo.

3. 🌐 **Diversidad de Modelos de Datos:** Existen varios tipos de bases de datos NoSQL, como las basadas en documentos, columnas, grafos y clave-valor. Cada una se adapta a diferentes tipos de datos y casos de uso.

4. 🔄 **Alta Disponibilidad y Tolerancia a Fallos:** Estas bases de datos suelen ser diseñadas para ofrecer alta disponibilidad y tolerancia a fallos, lo que garantiza que los datos estén siempre accesibles.

5. 📊 **Aplicaciones Diversas:** Las bases de datos NoSQL se utilizan en una amplia variedad de aplicaciones, desde redes sociales y aplicaciones web hasta IoT y análisis de datos.

**Ejemplos de Bases de Datos NoSQL:**

1. **MongoDB (Base de Datos de Documentos):** MongoDB es una base de datos NoSQL popular que almacena datos en documentos BSON. Por ejemplo, puedes tener una colección de documentos para usuarios, cada uno con campos diferentes según sus necesidades.

![](https://webimages.mongodb.com/_com_assets/cms/1-lwnlfl1ryn.png?auto=format%2Ccompress&ch=DPR)

2. **Cassandra (Base de Datos de Columnas):** Cassandra es una base de datos de columnas utilizada para grandes volúmenes de datos y alta escalabilidad. Es adecuada para aplicaciones como la gestión de registros de eventos.

![](https://www.instaclustr.com/wp-content/uploads/2021/10/Cassandra-Partitions-Partition-and-Clustering-Key.png)

3. **Neo4j (Base de Datos de Grafos):** Neo4j se utiliza para almacenar datos relacionales y su estructura se basa en nodos y relaciones. Es ideal para aplicaciones que requieren un análisis de grafos, como redes sociales.

![](https://miro.medium.com/v2/resize:fit:1400/1*dXf5fsV3siWk20lT8X5_ew.png)

4. **Redis (Base de Datos Clave-Valor):** Redis es una base de datos en memoria que almacena datos como pares clave-valor. Es ampliamente utilizado en cachés, colas de mensajes y aplicaciones en tiempo real.
![](https://redis.com/wp-content/uploads/2022/06/redis-enterprise-software-promo-image.svg?&auto=webp&quality=85,75&width=500)



**Ventajas de las Bases de Datos NoSQL:**

1. 🌟 **Flexibilidad de Esquema:** Las bases de datos NoSQL permiten cambios en la estructura de los datos sin requerir modificaciones en un esquema fijo. Esto es beneficioso para aplicaciones en evolución.

2. 🚀 **Escalabilidad Horizontal:** Son altamente escalables y permiten distribuir datos en múltiples servidores para manejar grandes volúmenes de información y cargas de trabajo crecientes.

3. 💻 **Manejo de Datos No Estructurados:** Pueden manejar datos no estructurados o semi-estructurados, como documentos JSON, que se ajustan bien a la web y aplicaciones modernas.

4. 🛡️ **Tolerancia a Fallos y Alta Disponibilidad:** Muchas bases de datos NoSQL ofrecen replicación automática y tolerancia a fallos, lo que garantiza la disponibilidad continua de datos.

5. 🔄 **Modelos de Datos Diversos:** Existen varios tipos de bases de datos NoSQL, lo que permite a los desarrolladores elegir el modelo de datos más adecuado para su aplicación (documentos, columnas, grafos, clave-valor).

**Desventajas de las Bases de Datos NoSQL:**

1. 🧩 **Complejidad de Consultas:** En comparación con las bases de datos relacionales, algunas bases de datos NoSQL pueden carecer de la capacidad para realizar consultas complejas o realizar agregaciones avanzadas.

2. 💾 **Falta de Transacciones ACID:** Muchas bases de datos NoSQL sacrifican la consistencia ACID (Atomicidad, Consistencia, Aislamiento, Durabilidad) en favor de un rendimiento y escalabilidad mejorados.

3. 🚧 **Menos Madurez y Ecosistema:** Algunas bases de datos NoSQL pueden carecer de una comunidad tan grande y madura como las bases de datos relacionales, lo que puede resultar en menos recursos y soporte disponibles.

4. 💼 **Falta de Estándares:** Los estándares en el mundo de las bases de datos NoSQL varían según el tipo y la implementación, lo que puede dificultar la migración entre sistemas.

5. 💳 **Costos Asociados:** Aunque inicialmente pueden ser más económicos en términos de hardware, a largo plazo, los costos operativos pueden aumentar debido a la necesidad de administrar sistemas altamente escalables.


## Mongodb

![](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRln12Czb3Uw73gzCaSO-FSvLEeuQ_Ikr67lKtUmfX3k7aZT5MBtnZUa6B5isewEJXorg&usqp=CAU)


**Historia y Evolución de MongoDB:**

🌱 **Orígenes:** MongoDB fue desarrollado por la compañía 10gen Inc. (ahora MongoDB Inc.) y se lanzó por primera vez en 2009. El proyecto se inspiró en sistemas de base de datos orientados a documentos como CouchDB y Lotus Notes.

🔄 **Versiones Principales:** A lo largo de los años, MongoDB ha experimentado múltiples versiones y actualizaciones importantes. Algunas de las versiones más notables incluyen MongoDB 1.0 en 2009, MongoDB 2.2 en 2012 con características de agregación, MongoDB 3.0 en 2015 con almacenamiento en cable y MongoDB 4.0 en 2018 con transacciones ACID.

👥 **Community y Enterprise Edition:** MongoDB está disponible en dos ediciones: Community Edition, que es de código abierto y gratuita, y Enterprise Edition, que ofrece características adicionales y soporte comercial. Esta dualidad ha permitido su adopción tanto en proyectos de código abierto como en empresas.

☁️ **Ecosistema:** MongoDB ha desarrollado un rico ecosistema de herramientas y servicios relacionados, como MongoDB Atlas, una plataforma de bases de datos en la nube, y el lenguaje de consulta Aggregation Framework, que permite consultas más avanzadas.

**Características Principales de MongoDB:**

1. 📄 **Modelo de Datos de Documentos:** MongoDB almacena datos en documentos BSON (Binary JSON), lo que permite una estructura flexible y anidada. Esto facilita la representación de datos semiestructurados.

2. 🚀 **Escalabilidad Horizontal:** MongoDB es altamente escalable, lo que significa que puedes distribuir tu base de datos en múltiples servidores para manejar grandes volúmenes de datos y cargas de trabajo.

3. 🔍 **Consultas y Agregaciones Poderosas:** MongoDB ofrece un lenguaje de consulta avanzado y un conjunto de herramientas para realizar consultas, agregaciones y análisis de datos.

4. 🛡️ **Alta Disponibilidad y Tolerancia a Fallos:** MongoDB admite replicación automática y tolerancia a fallos, lo que garantiza que los datos estén siempre disponibles.

5. 📑 **Índices y Text Search:** Ofrece capacidades de indexación para mejorar la velocidad de las consultas y también admite búsqueda de texto completo.

6. 💼 **Transacciones ACID:** A partir de la versión 4.0, MongoDB ofrece transacciones ACID para garantizar la consistencia de los datos en aplicaciones empresariales.

7. 🌐 **Soporte para Múltiples Plataformas:** MongoDB se ejecuta en una variedad de sistemas operativos y plataformas, lo que lo hace versátil y adecuado para una amplia gama de casos de uso.



### MongoDb

Para probar MongoDB Atlas, el servicio de base de datos en la nube de MongoDB, sigue estos pasos para obtener una cuenta gratuita y comenzar a usarlo:

1. **Visita el Sitio Web de MongoDB:**
   Abre tu navegador web y visita el sitio web oficial de MongoDB en https://www.mongodb.com/es.

2. **Clic en "Comienza Gratis":**
   En la página de inicio, es probable que encuentres un botón o enlace que diga "Comienza Gratis" o algo similar. Haz clic en ese enlace.

3. **Crea una Cuenta o Inicia Sesión:**
   Si aún no tienes una cuenta de MongoDB, te pedirán que la crees. De lo contrario, inicia sesión con tus credenciales existentes.

4. **Completa el Formulario de Registro:**
   Si estás creando una cuenta, deberás proporcionar información como tu nombre, dirección de correo electrónico y contraseña. Rellena el formulario y sigue las instrucciones.

5. **Verifica tu Correo Electrónico:**
   Después de registrarte, es posible que debas verificar tu dirección de correo electrónico. Revisa tu bandeja de entrada para un correo de verificación de MongoDB y sigue las instrucciones para confirmar tu dirección de correo.

6. **Comienza con MongoDB Atlas:**
   Una vez que hayas completado el registro y verificado tu dirección de correo, inicia sesión en tu cuenta. Busca opciones relacionadas con MongoDB Atlas o "Crear un Proyecto" en tu panel de control.

7. **Configura tu Proyecto en MongoDB Atlas:**
   Sigue las instrucciones para configurar un nuevo proyecto en MongoDB Atlas. Esto implicará seleccionar una nube, una región, configurar opciones de seguridad, etc.

8. **Crea un Cluster Gratuito:**
   Durante la configuración de tu proyecto, podrás crear un clúster gratuito. Selecciona la opción gratuita o de prueba, y sigue las instrucciones para lanzar tu clúster en la nube.

9. **Conéctate a tu Clúster:**
   Una vez que el clúster esté en funcionamiento, podrás conectarte a través de la cadena de conexión proporcionada, ya sea a través de la línea de comandos o una herramienta de MongoDB.

Recuerda que MongoDB Atlas ofrece una capa gratuita con ciertas restricciones de uso. A medida que tu proyecto crezca, es posible que desees considerar las opciones de actualización y pago según tus necesidades. ¡Disfruta explorando MongoDB con MongoDB Atlas!


En Mongo, los conceptos claves son:
- los <b>Documentos</b>: un documento corresponde a objeto JSON.
- las <b>Colecciones</b>: una colección es un conjunto de documentos. Nota Bene: en una colección Mongo, los documentos pueden tener una estructura distinta.

En Mongo, no existen directamente los conceptos de Tabla, Tupla o Claves primerias/foraneas. Sin embargo, podemos ver una equivalencia entre Tabla y Colección, y entre Tupla y Documento.

Verificar que Mongo está instalado correctamente ejecutando:
- el programa principal del SGBD: <code>mongod &</code> desde un terminal
- el programa CLI (Command Line Interface): <code>mongo</code>

## Utilizar Mongo desde el Terminal

Seguir el Tutorial "Mongo" de TutorialsPoint.com: https://www.tutorialspoint.com/mongodb/index.htm

## Utilizar Mongo desde un entorno Python

Vamos a interactuar con Mongo desde un entorno Python, por lo tanto utilizaremos el driver <code>PyMongo</code> (que instalar).

In [58]:
pip install pymongo

In [59]:
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi
uri = "mongodb+srv://BIG_DATA_EJEMPLO:TodyFtzHKBOw886d@cluster0.nmkhay1.mongodb.net/?retryWrites=true&w=majority"
# Create a new client and connect to the server
client = MongoClient(uri, server_api=ServerApi('1'))
# Send a ping to confirm a successful connection
try:
    client.admin.command('ping')
    print("Pinged your deployment. You successfully connected to MongoDB!")
except Exception as e:
    print(e)

Pinged your deployment. You successfully connected to MongoDB!


Ver información de la conexión

In [60]:
client.stats

Database(MongoClient(host=['ac-c90zkja-shard-00-01.nmkhay1.mongodb.net:27017', 'ac-c90zkja-shard-00-02.nmkhay1.mongodb.net:27017', 'ac-c90zkja-shard-00-00.nmkhay1.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-oyrhi3-shard-0', tls=True, server_api=<pymongo.server_api.ServerApi object at 0x794fa4603490>), 'stats')



- `Database`: Esto indica que se trata de una instancia de una base de datos en MongoDB.

- `MongoClient`: Indica que se ha utilizado el cliente MongoDB para establecer una conexión con un conjunto de servidores de MongoDB. El cliente MongoDB es responsable de la conexión a la base de datos.

- `host`: Aquí se especifica la información de los servidores de MongoDB a los que se ha conectado. La conexión se realiza a través de una lista de hosts que forman parte de un conjunto de réplicas (replica set) en MongoDB. Cada servidor está identificado por su dirección IP o nombre de host y número de puerto (en este caso, el puerto 27017).

- `document_class`: Esto parece indicar el tipo de clase que se utiliza para representar documentos en la base de datos. En este caso, se está utilizando un diccionario (dict) para representar los documentos.

- `tz_aware`: Indica si la base de datos es consciente de la zona horaria (timezone-aware). En este caso, se establece en `False`, lo que significa que la base de datos no tiene en cuenta las zonas horarias.

- `connect`: Indica si la conexión a la base de datos está activada. En este caso, está configurado como `True`, lo que significa que la conexión está habilitada.

- `retrywrites`: Este parámetro controla si se deben volver a intentar las escrituras en caso de error. Está configurado como `True`, lo que significa que las escrituras se reintentarán en caso de error.

- `w`: Indica el nivel de consistencia de escritura. En este caso, está configurado como 'majority', lo lo que significa que se espera que la mayoría de los miembros del conjunto de réplicas confirmen las escrituras.

- `authsource`: Indica la base de datos de autenticación. En este caso, se establece como 'admin', lo que significa que la autenticación se realiza en la base de datos "admin".

- `replicaset`: Hace referencia al nombre del conjunto de réplicas al que se ha conectado.

- `tls`: Indica si se está utilizando TLS (Transport Layer Security) para la conexión segura a MongoDB. En este caso, está configurado como `True`, lo que significa que se está utilizando TLS.

- `server_api`: Puede estar relacionado con la configuración de una versión específica de la API del servidor MongoDB.


In [61]:
print(client.list_database_names())

['ejemplo', 'sample_airbnb', 'sample_analytics', 'sample_geospatial', 'sample_guides', 'sample_mflix', 'sample_restaurants', 'sample_supplies', 'sample_training', 'sample_weatherdata', 'admin', 'local']


La información de las bases de datos disponibles se encuentra en [aquí](https://www.mongodb.com/docs/atlas/sample-data/)

**Establece el nombre de la base de datos con la que trabajar. Si no existe, se creará tan pronto como se añada un documento.**


In [62]:
db=client.sample_airbnb

Revisemos las colecciones asociadas a la base de datos

In [63]:
db.list_collection_names()

['listingsAndReviews']

Para acceder a la colección

In [64]:
coleccion=db.listingsAndReviews
coleccion

Collection(Database(MongoClient(host=['ac-c90zkja-shard-00-01.nmkhay1.mongodb.net:27017', 'ac-c90zkja-shard-00-02.nmkhay1.mongodb.net:27017', 'ac-c90zkja-shard-00-00.nmkhay1.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-oyrhi3-shard-0', tls=True, server_api=<pymongo.server_api.ServerApi object at 0x794fa4603490>), 'sample_airbnb'), 'listingsAndReviews')

In [65]:
documentos=coleccion.find()
for documento in documentos[0:5]:
    print(documento)

{'_id': '10006546', 'listing_url': 'https://www.airbnb.com/rooms/10006546', 'name': 'Ribeira Charming Duplex', 'summary': 'Fantastic duplex apartment with three bedrooms, located in the historic area of Porto, Ribeira (Cube) - UNESCO World Heritage Site. Centenary building fully rehabilitated, without losing their original character.', 'space': 'Privileged views of the Douro River and Ribeira square, our apartment offers the perfect conditions to discover the history and the charm of Porto. Apartment comfortable, charming, romantic and cozy in the heart of Ribeira. Within walking distance of all the most emblematic places of the city of Porto. The apartment is fully equipped to host 8 people, with cooker, oven, washing machine, dishwasher, microwave, coffee machine (Nespresso) and kettle. The apartment is located in a very typical area of the city that allows to cross with the most picturesque population of the city, welcoming, genuine and happy people that fills the streets with his o

Revisión de claves

In [66]:
list(db.listingsAndReviews.find_one())

['_id',
 'listing_url',
 'name',
 'summary',
 'space',
 'description',
 'neighborhood_overview',
 'notes',
 'transit',
 'access',
 'interaction',
 'house_rules',
 'property_type',
 'room_type',
 'bed_type',
 'minimum_nights',
 'maximum_nights',
 'cancellation_policy',
 'last_scraped',
 'calendar_last_scraped',
 'first_review',
 'last_review',
 'accommodates',
 'bedrooms',
 'beds',
 'number_of_reviews',
 'bathrooms',
 'amenities',
 'price',
 'security_deposit',
 'cleaning_fee',
 'extra_people',
 'guests_included',
 'images',
 'host',
 'address',
 'availability',
 'review_scores',
 'reviews']

In [67]:
db = client.ejemplo
client.list_database_names()

['ejemplo',
 'sample_airbnb',
 'sample_analytics',
 'sample_geospatial',
 'sample_guides',
 'sample_mflix',
 'sample_restaurants',
 'sample_supplies',
 'sample_training',
 'sample_weatherdata',
 'admin',
 'local']

In [68]:
db.create_collection('adressbook')

CollectionInvalid: ignored

In [69]:
list (db.list_collections())

[{'name': 'adressbook',
  'type': 'collection',
  'options': {},
  'info': {'readOnly': False,
   'uuid': Binary(b'\x1f\xedw3\x0c\x17HK\xac\xf1{-?WTO', 4)},
  'idIndex': {'v': 2, 'key': {'_id': 1}, 'name': '_id_'}}]

In [70]:
# Set the collection to work with
collection = db.adressbook
collection.insert_one({'name' : 'Antonino'})

InsertOneResult(ObjectId('65444ff36d9415308f4882f3'), acknowledged=True)

In [71]:
list (collection.find())

[{'_id': ObjectId('65444a466d9415308f4882ee'),
  'name': 'Antonino',
  'age': 85,
  'gender': 'Male',
  'likes_python': False,
  'address': {'street': "Torrent de l'Ollafghfjfgj",
   'number': 704,
   'city': 'Barcelonacvn',
   'floor': None,
   'postalcode': '080124'},
  'favouriteFruits': ['banana', 'pineapple', 'orange']},
 {'_id': ObjectId('65444ff36d9415308f4882f3'), 'name': 'Antonino'}]

In [ ]:
# Renombrar una colección
db.adressbook.rename('agenda')
colección = db.agenda  # Establecer la colección con la que trabajar
list(db.list_collections())  # Mostrar las colecciones existentes

# Eliminar una colección
db.drop_collection('agenda')


**Operaciones Básicas en Colecciones**

Operadores de Consulta - Crear, Leer, Actualizar y Eliminar (un documento)


Crear un documento con: insert_one()


Para insertar un documento se puede hacer usando: insert_one y notación  { 'nombreDelAtributo1' : 'contenido1', 'nombreDelAtributo2' : 'contenido2', ... }

In [ ]:
#@title Contact

Name = "Antoninoaefef" #@param {type:"string"}
Age = 41 #@param {type:"slider", min:10, max:80, step:1}
Gender = "Female" #@param ["Male", "Female"]
Likes_Python = "Yes" #@param ["Yes", "No"]
if Likes_Python is "Yes":
  Likes_Python = True
else:
  Likes_Python = False


#@markdown Address
Street = "Torrentfghfjfgj" #@param {type:"string"}
Number = 70 #@param {type:"integer"}
City = "Barcelonacvn" #@param {type:"string"}
PostalCode = "080124" #@param {type:"string"}

In [ ]:
data = {  'name' : Name ,                                    # String
          'age' : Age,                                       # Integer
          'gender' : Gender,                                 # String
          'likes_python' : Likes_Python,                     # Boolean
          'address': {
              'street' : Street,                             # String ( special character with escape \ )
              'number' : Number,                             # Integer
              'city' : City,                                 # String
              'floor' : None,                                # Null
              'postalcode' : PostalCode,                     # String containing a number
              },
          'favouriteFruits': ['banana','pineapple','orange'] # Array
       }

insert_result = collection.insert_one( data)

In [ ]:
insert_result.acknowledged

In [ ]:
insert_result.inserted_id

**Leer documento  con: find()**

Para leer un documento se puede hacer utilizando find y notación en JavaScript { 'atributo1 que contiene la clave' : 'clave1 a encontrar' }

In [ ]:
list ( collection.find() )

In [ ]:
list ( collection.find( {'_id' : insert_result.inserted_id } ))

In [ ]:
list ( collection.find( {'name' : 'Antonino' } ))

In [ ]:
list ( collection.find( {'address.city' : City } ))

In [ ]:
list ( collection.find().limit(1) )

In [ ]:
list ( collection.find().skip(1) )

**Actualizar un documento con: update_one()**

In [ ]:
Name

In [ ]:
update_result = collection.update_one(
    {'name' : Name},
    {'$set' : { 'age' : 30 }} )

list (collection.find( {'name' : Name } ))

In [ ]:
update_result.raw_result

In [ ]:
insert_result = collection.update_one( {'name' : 'Juan Gomez'}, {'$set' : { 'age' : 85 }}, upsert= True )
list (collection.find( {'name' : 'Juan Gomez'} ))

In [ ]:
list(collection.find().limit(10))

**Eliminar un documento con: delete_one()**



In [ ]:
delete = collection.delete_one({'name': 'Juan Gomez'})

In [ ]:
delete.deleted_count   # informs that 1 document has been deleted

In [ ]:
collection.delete_one({'name': Name})

### Crear, Leer, Actualizar, Eliminar (muchos documentos)

Crear un documento con: insert_many()
Para insertar varios documentos se puede hacer utilizando insert_many y notación  { 'atributoDocumento1' : 'contenido1' }, { 'atributoDocumento2' : 'contenido2' }, { ... }

In [ ]:
import datetime
collection.insert_many(  [                          # <---- start a list with [
##  Insert Document 1
  {
  'name': 'Jordi Gonzalez',
  'age': 25,
  'likes_python': True,
  'registered': datetime.datetime(2015, 2, 11, 4, 22, 39),
  'address': {
      'street': 'Torrent de l\'Olla',
      'number': 70,
      'floor': None,
      'city': 'Barcelona',
      'postalCode': '08012'
             },
  'height':  1.72,
  'favouriteFruits': ['banana','pineapple','orange']
  },

##  Insert Document 2
  {
  'name': 'Maria Smith',
  'age': 30,
  'likes_python': True,
  'registered': datetime.datetime(2016, 4, 23, 7, 34, 12),
  'address': {
      'street': 'Numancia',
                                                     ##  missing number
                                                     ##  missing floor
      'city': 'Barcelona',
      'postalCode': '08029'
             },
  'height':  1.56,
  'favouriteFruits': ['lemon','pineapple']
  }
  ]   )                                                   # <---- finalize the list ]

### Leer varios documentos con find():
"Para leer un documento se puede hacer utilizando `find` y notación  { 'atributo1 que contiene la clave' : 'clave1 a encontrar' }"

In [ ]:
list ( collection.find( {'$or': [ {'name': 'Jordi Gonzalez'},{'name': 'Maria Smith'} ]}))

 **Actualizar varios documentos con: update_many()**

In [ ]:
collection.update_many( {'isActive': True }, {'$set' : { 'isActive': False }} )
list (collection.find( ))

**Borrar varios documentos con: delete_many()**

In [ ]:
delete = collection.delete_many({'likes_python': True})
list (collection.find( ))

In [ ]:
delete.deleted_count

# Example 1 - Bicing stations

In [73]:
!pip install folium

In [74]:
import pymongo                            # Library to access MongoDB
from pymongo import MongoClient           # Imports MongoClient
import pandas as pd                       # Library to work with dataframes
import folium                             # Library to visualize a map

In [ ]:
# Definir la URI de conexión
uri = "mongodb+srv://Zainea:4JODSMLaVrXziVK6@clustermongo.2vcbgpa.mongodb.net/test"

# Conectar a la base de datos
client = MongoClient(uri)





In [ ]:

# Seleccionar la base de datos y la colección
db = client.brny4kjelauboxl
collection = db.bicing

# Cargar el conjunto de datos desde un archivo CSV
df = pd.read_csv("/content/bicing_data.csv")  # Reemplaza "ruta_del_archivo.csv" con la ruta de tu archivo CSV

# Convertir el DataFrame a un formato adecuado para MongoDB (lista de diccionarios)
data = df.to_dict(orient="records")

# Insertar los datos en la colección
collection.insert_many(data)

##3. Connect to database

In [ ]:
db = client.brny4kjelauboxl               # Set the database to work on
db.list_collection_names()                # List the collections available
collection = db.bicing                    # Collection alias

##4. Quick data overview

In [ ]:
num_documents = collection.count_documents({'_id' : {'$exists' : 1}})
print ( 'Number of documents in database = ' + str(num_documents) )
list ( collection.find().limit(1) )

In [ ]:
# The values of 'bikes' is string type instead of number.
# In order to filter by number greater than, we need to convert the value to integer.
# to change an attribute type, can be done easily with mongoDB 4
# In workshop we use mongo 3, I suggest to use the following method to convert it

bikes_list = list(collection.distinct('bikes'))             # list the unique values of 'bikes', we get a list of strings
for num in bikes_list:                                      # iterate the list, item by item
  collection.update_many({'bikes' : num},{'$set': {'bikes' : int(num)}})    # update each document with a number in string with the same number as Integer


Consulta a la base de datos: Obtener estaciones activas con al menos 3 bicicletas.

In [ ]:
# Loading database query in pandas Dataframe
filters = {'status':'OPN', 'bikes' : {'$gte' : 3 }}   # Usage of gte Query Operator  $gte = "greater than or equal"
fields = { '_id', 'latitude' , 'longitude', 'bikes', 'slots'}

query = list( collection.find( filters , fields ) )
df = pd.DataFrame ( query )                             # Load the database reply in a Pandas DataFrame

In [ ]:
print ( 'Numer of active stations with at least 3 bicycles: ' + str(len (query)) )

In [ ]:
df.iloc[0] # prints the first DataFrame row

##6. Mark Bicing stations in map

In [ ]:
center_lat = 41.378
center_lon = 2.139

locationmap = folium.Map(location=[ center_lat , center_lon ], zoom_start=16, width=800, height=600 )
longitud  = len( df )

for i in range ( longitud ):
    lng = float(df.iloc[i]['longitude'])
    lat = float(df.iloc[i]['latitude'])
    description = 'Bikes: ' + str(df.iloc[i]['bikes']) + '<br> Empty slots: ' + str(df.iloc[i]['slots'])
    folium.Marker( [ lat , lng ],
                 popup= description,
                 icon=folium.Icon(color='red')).add_to(locationmap)

locationmap